### ISSUE TITLE
Research about the implementation

### ISSUE THREAD

**Issue by nik-55** [nik-55/hingli#1]
> Initial Proposal

# Overview

A python package that deals with ML operations on Hinglish. The package will offer the  following functionalities:

1. Convert Hinglish to English.
2. Extract emotions (joy, anger, sadness, depression) from the Hinglish text.
3. Check if Hinglish text is abusive.

## Use Case

Millions of Indians communicate in Hinglish on different platforms like social media. Developers can use this package to achieve the following objectives easily:

1. Extracting emotions from Hinglish tweets, comments, captions, reviews, and messages will help to know people’s sentiments about particular stocks, companies or products, government decisions, and political leaders.  
2. Delete comments or messages in public forums that contain abusive words.
3. Provide an option to translate Hinglish messages in open communities to English. 

# Technical Implementation

## Emotion detection from Hinglish text

Text-based emotion detection is based on the assumption that when a person is happy, they would use positive words. Likewise, when they are angry, frustrated or upset, negative emotions will be depicted by a certain kind of words carrying negative connotations.

This is a multi-label text classification problem.

### Dataset

1. [Dataset Mentioned in Research Paper](https://github.com/anshulwadhawan/emotion_detection/tree/master/Datasets)- The dataset contains the tweet ID for 150K Hinglish tweets, and each tweet has been labeled with six emotional classes as follows:

![image](https://user-images.githubusercontent.com/98151272/233711652-2b2b3b53-05ef-4bf9-b32e-a005f38c7250.png)
     
[Twitter API](https://developer.twitter.com/en/docs/twitter-api/tweets/lookup/api-reference/get-tweets-id) can be used to get tweet content from the tweet ID present in the dataset.

2. [Download SAIL_2017_DATASET](http://amitavadas.com/SAIL/Data/SAIL_2017.zip)- The dataset contains the Hinglish texts, and each text is labeled with positive, negative, and neutral sentiments. However, this dataset does not provide the classification based on emotions.

### Text Preparation

1. Emoji to UTF-8 Encoding.
2. Removing Stop words.
3. Word Tokenization.

### Feature Engineering

Word embeddings are vector representations of words. Word2vec can be utilized due to the fact that words that exhibit semantic similarity also tend to have similar vector representations.

### DL Model

Long Short-Term Memory (LSTM) is effective in capturing long-term dependencies in sequential data, making it a popular choice for text classification tasks. Bi-LSTM is an extension of LSTM that allows the model to take into account the context of the input sequence in both forward and backward directions. Bi-LSTM-based text classification models have outperformed traditional approaches in sentiment analysis. 

### References

1. [Article- Emotion Detection in Hinglish(Hindi+English) Code-Mixed Social Media Text](https://www.sciencedirect.com/science/article/pii/S1877050920311236)
2. [Project- Hinglish-Sentiment-Analysis](https://github.com/vipul-khatana/Hinglish-Sentiment-Analysis)
3. [Research Paper](https://aclanthology.org/2021.wassa-1.21.pdf)

Ps: Pre-trained models can also be utilized. However, it mostly depends on the availability and suitability to this task.

## Hinglish to English translation

Pre-trained models can be useful for Hinglish to English translation tasks, as they are already trained on a large amount of text data and can capture complex linguistic relationships between words in a sentence.

1. [Research Paper](https://aclanthology.org/2021.ranlp-srw.3.pdf)- It mentions about using the mBART and mT5 based machine translation model for hinglish to english translation.
    * mBART is a powerful pre-trained model that can be used for a variety of multilingual NLP tasks, including Hinglish to English translation. Its ability to handle multiple languages and transfer knowledge between them makes it particularly useful for multilingual applications.
    * [Dataset](https://aclanthology.org/D18-1076/)- It is used to finetune and test the model. It consists of roughly 10K English and Hinglish codemixed sentences.

## Abuse Detection

1. [Research Paper](https://arxiv.org/pdf/1809.08652v1.pdf)- It mentions about using LSTM based classifier model. This model classifies on three classes- non offensive, hate inducing and abusive. 
2. Project: [Hinglish-hate-detection](https://github.com/victor7246/Hinglish_Hate_Detection)
    * [Dataset](https://github.com/victor7246/Hinglish_Hate_Detection/tree/main/data)- It contains text that are labeled with yes or no depending on whether it is offensive or not.

## Python Package

Steps for building a python package can be found [here](https://www.freecodecamp.org/news/build-your-first-python-package/).

### Folder Structure

![image](https://user-images.githubusercontent.com/98151272/233711471-0c72ce16-3dbc-4dd5-b22b-1496946f600f.png)
